eth2/utils/merkle_partial/src/partial.rs

```diff
@@ -0,0 +1,426 @@
+
+                cache.insert(parent, h);
+                nodes.push(parent);
+            }
+
+            position += 1;
+        }
+
+        Ok(())
+    }
+
+    /// Recursively traverse the tree structure, matching the appropriate `path` element with its index,
+    /// eventually returning the `indicies` and `chunks` needed to generate the partial for the path.
+    fn get_partial_helper(
+        &self,
+        cache: &Cache,
+        root: NodeIndex,
+        height: usize,
+        path: Vec<&str>,
+        indices: &mut Vec<NodeIndex>,
+        chunks: &mut Vec<u8>,
+    ) -> Result<(Vec<NodeIndex>, Vec<u8>)> {
+        if path.len() == 0 {
+            return Ok((indices.clone(), chunks.clone()));
+        }
+
+        let path_element = path[0];
+
+        let mut leaves: Vec<Node> = vec![];
+        for i in 2_u32.pow(height as u32)..2_u32.pow(height as u32 + 1) {
+            leaves.push(self.get_node(i as NodeIndex));
+        }
+
+        for leaf in leaves {
+            if let Node::Leaf(Leaf::Basic(chunk_fields)) = leaf {
+                for field in chunk_fields {
+                    if path_element == field.ident {
+                        let index = field.index;
+
+                        indices.push(index);
+                        chunks.extend(cache.get(index).ok_or(Error::MissingNode(index))?);
+
+                        let mut visitor = index;
+
+                        while visitor > root {
+                            if let Some(sibling) = get_sibling_index(visitor) {
+                                indices.push(sibling);
+                                chunks
+                                    .extend(cache.get(sibling).ok_or(Error::MissingNode(sibling))?);
+                            }
+
+                            visitor /= 2;
+                        }
+
+                        // should recurse here for container types
+                        return Ok((indices.clone(), chunks.clone()));
+                    }
+                }
+            }
+        }
+
+        Err(Error::InvalidPath(path_element.to_string()))
+    }
+}
+
+/// Helper function that appends `right` to `left` and hashes the result.
+fn hash_children(left: &[u8], right: &[u8]) -> Vec<u8> {
+    let children: Vec<u8> = left.iter().cloned().chain(right.iter().cloned()).collect();
+    hash(&children)
+}
+
+/// Returns the index of a node's sibling.
+fn get_sibling_index(index: NodeIndex) -> Option<NodeIndex> {
+    if index == 0 {
+        return None;
+    }
+
+    if index % 2 == 0 {
+        Some(index + 1)
+    } else {
+        Some(index - 1)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::super::vec_to_array;
+    use super::*;
+    use crate::field::{Basic, Leaf, Node, Value};
+    use ethereum_types::U256;
+    use std::mem::transmute;
+
+    #[derive(Debug, Default)]
+    struct A {
+        a: U256,
+        b: U256,
+        c: u128,
+        d: u128,
+        // Should be added by derive macro
```

Comment:
I'm not certain this is true, but I don't believe you can add a field to a struct with a derive macro.

Reply:
You're totally right!! 😄